### ISSUE TITLE
wrong order while doing mean subtraction

### ISSUE THREAD

**Issue by vuhonganh** [yhenon/keras-frcnn#148]
I think there is a mistake when you subtract the mean channel-wise. In config.py, the order of the image channel-wise mean is already BGR (the author of Resnet stated here #https://github.com/KaimingHe/deep-residual-networks/issues/5). In addition, you use opencv to load image, hence it also load by BGR order (please see [here](http://docs.opencv.org/3.0-beta/doc/py_tutorials/py_gui/py_image_display/py_image_display.html)). In data_generators.py, you transpose BGR -> RGB and then apply mean subtraction but this is not correct since they are already in good order (BGR). 
By the way, can you share your thoughts about the std_scaling, classifier_regr_std? I really don't get where it comes from and why we need it.

**Comment by liuwei16** [comment]
I also find this error, thus 
x_img = x_img[:,:, (2, 1, 0)]  # BGR -> RGB
should be commented.

**Comment by liuwei16** [comment]
In caffe, the input image is ordered as [batchsize channels height width].
but in tensorflow the order is  [batchsize height width channels ].
Can you still use the weights from the pretrained model?

**Comment by vuhonganh** [comment]
@liuwei16 yes we can, this is done by separating the backend case. Keras supports 2 backend: Theano (th), and Tensorflow (tf). You can see that @yhenon deal with it by considering backend K.dim_ordering.

**Comment by liuwei16** [comment]
@vuhonganh @yhenon Also, I have some confusions about the regression part.
1、 C.std_scaling = 4.0, classifier_regr_std=[8,8,4,4], where is this consideration from? Is it the same in the caffe version?
2、I found the anchor regression is operated on the downscaled grid cells, the regression result is in the type of int, is this harmful for tiny objects with width or height less than 16 pixels?
Thank you very much

**Comment by giladdiv** [comment]
@vuhonganh so you claim that the network is using RGB image or BGR?
because as i see it now, you load BGR image -> convert to RGB->substract mean ->convert to BGR and input it to the net. what is the problem?

**Comment by vuhonganh** [comment]
@giladdiv i think you should review my first comment to understand, it's clear from the first sentence: "I think there is a mistake when you subtract the mean channel-wise."

**Comment by anandkoirala** [comment]
@vuhonganh 
is the code all right with the mean subtraction or needs changing?
@liuwei16 
did you find what C.std_scaling=4.0, classifier_reg_std=[8,8,4,4] doing? i think it is to do with zero mean and unit variance but where these values came from and why is scaling done?
Yes if object size is less than 16 pixels the model will struggle for localization and small objects more error because of the IoU implementation. It is better to resize or split your image such that the size of smallest object is greater than 16 pix... thanks

**Comment by JieZou1** [comment]
@anandkoirala @liuwei16 

Any further ideas about C.std_scaling. What is its purpose? Also, for small objects less than 16x16 pixels, why would IoU implementation affects localization? If we enlarge the image, it is going to significantly increase the computation time, right?
